Improved Full Detector Response storage and conversion

jdbuhler opened the pull request:
This PR overhauls the on-disk and in-memory format of the FullDetectorResponse. The principal goals are to

 * reduce on-disk storage size via HDF5 internal compression, without incurring an unacceptable running time penalty
 * reduce rsp->h5 conversion time and memory usage
 * allow h5->rsp conversion without data loss

The biggest change is that the raw CDS ring counts are stored as small integers on disk, separately from the effective area multipliers.  This change greatly improves compressibility and reduces peak memory during rsp->h5 conversion.  As a bonus, it allows more efficient weighted averaging of slices at runtime and dynamic selection of 32- vs 64-bit PSRs.

Conversion has been separated from the FullDetectorResponse class into its own RspConverter class, which handles both rsp->h5 and h5->rsp.

After consultation with Israel et al., these changes remove support for the following deprecated formats and features:
 * sparse responses
 * miniDC2 format
 * reading the spectrum for effective area computation from a file (was already broken in DC3 codebase)

The new on-disk response format is *not* backwards or forwards compatible with the previous format.  For this reason, I've established a new directory COSI-SMEX/develop in the public cosipy bucket on wasabi, put the new response files there, and updated the tutorials to use them instead of the ones from DC2/DC3.  It is anticipated that DC4 will use the new format, while DC3 will remain with the current one.

The patch introduces a dependency on the hdf5plugin package, a well-supported companion to h5py that implements the BitShuffle compression algorithm used by the new response format.

## Supporting performance numbers

### Conversion time and memory

DC2 O3 continuum response:  < 4 minutes, 4 GB
DC3 O3 polarization response:  < 6 minutes, 6.33 GB
O4 version of continuum response: < 90 minutes, 52.7 GB

Testing was on a 2.3 GHz Intel Xeon Gold 5118 server with 192 GB of RAM and SSD storage; only one core was used.

(The first two used to take hours, while the last took overnight and needed at least half a terabyte of RAM)

### HDF5 response file sizes on disk

DC2 O3 continuum response: 592.76 MB  (.rsp.gz is 841.58 MB)
DC3 O3 polarization response: 795.76 MB (.rsp.gz is 954.58 MB )
O4 version of continuum response: 4187.05 MB (.rsp.gz is 5186.87 MB)

These file sizes are < 1/10 the size of the previous uncompressed HDF5 responses for O3, and more like 1/50 for O4
response, assuming (as was the case for the file I received) the latter is stored in float32 precision.

### Read times

The following are average times to read one slice from the NuLambda axis (i.e., rsp[i]).  Testing was on the same machine used for the conversion tests.

DC2 O3 continuum response: 4.5 ms (old), 6.4 ms (new)
DC3 O3 polarization response:  8.8 ms (old), 11 ms (new)
O4 version of continuum response: 350 ms (old), 46 ms (new)

"Old" is the previous response format, while "new" is the format in this PR.  Note that for large enough responses, such as the O4 continuum response, it's faster to read less data from disk and decompress than to read the data uncompressed.

By way of comparison, compressing the DC2 O3 continuum response with HDF5's gzip internal compression yielded a substantially *larger* file (889.77 MB) and read times of 21 ms, > 3x longer than with the BitShuffle compression used in this patch.

Commit:
* Break out rsp.gz to hdf5 conversion from full data response
  into a new RspConverter class
* change on-disk response format to split out counts from
  effective area, and use low-overhead compression on counts
* allow counts to be stored in integer types smaller than 32 bits;
  provide auto-detection of the best size as an option
* Rework .rsp conversion code to use less memory
* Remove outdated support for sparse .rsp and broken code for
  file-based normalization
* Enable axes of response to be read and written via the Axes object
  instead of replicating its functionality.  Keep textual descriptions
  of axes in a separate group for pretty-printing.
* Allow each Healpix axis to have its own nside; all-sky axes use
  nside=1
* capture and store header fields other than the axes and the
  size of counts in the .h5 file for future reference

Commit:
* Make more of the behavior of RspCoverter driven by
tables stored at class scope.
* Filter axes to be used for output HDF5 file as they
are being read from .rsp, rather than arbitrarily
removing the last couple of axes afterwards.

Commit:
* replace explicit use of guts of response with an API call in image deconvolution
* make default for response construction to infer the element size from the
  data, rather than guess a too-large size

Commit:
* remove obsolete miniDC2 support which made strong assumptions about
  the content of the response
* fix __array__ method of FullDetectorResponse
* use new test full detector and polarization responses, and update
  test suite to expect the outputs that occur when they are used
  (work in progress)

Commit:
* update some more tests with output for new response
* remove backup file that snuck into docs

Commit:
* don't exchange Phi, PsiChi axes -- test suite passes without it
* add conversion to Histogram to FullDetectorResponse API, rather
  than open-coding it in image_deconvolution/dataIF

Commit:
turn off resampling of covariance matrix in 3ML fitting, which is
not relevant to the output being tested and occasionally causes crashes

Commit:
simplify some tests that explicitly check the energy edge array and nside of an axis

Commit:
* add functionality to write .rsp.gz file from an open FullDetectorResponse
* add unit tests for response conversion between .rsp and .h5
* fix typo in FullDetectorResponse.to_histogram()

Commit:
* add method to extract effective area from response
* remove old commented-out code

Commit:
* eff_area cannot be both local field and exposed property
* fix typo in open()

Commit:
* add versioning to HDF5 response, and do version check when opening
* to further avoid confusion between response versions, rename
  CONTENTS dataset to COUNTS, since it is now raw counts
* FullDetectorResponse.open() should support only HDF5, as we cannot
  use an .rsp file live, it is not reasonable to convert a usefully
  large (o3 or above) .rsp.gz file "on the fly", and we have a
  separate converter class now.

Commit:
* Do not use HDF5's timestamps or order-tracking features, so that converting the same
.rsp to .h5 twice yields identical byte streams.
* Do order tracking of headers ourselves instead of relying on HDF5 to do it.

Commit:
* Update notebooks to point to new-format responses in "develop" rather than "DC2/DC3",
  with new checksums.  Shorten the names while we're at it.
* Remove vestiges of miniDC2 from image deconvolution notebooks
* make sure setup for cosipy installs hdf5plugin package

Commit:
document key features of new response format and
existence of COSI-SMEX/develop tree on wasabi

Commit:
* add source for test polarization response to test data
* update .h5s for test full detector responses to newest on-disk format

Commit:
Add option to FullDetectorResponse.open() to set the datatype of
eff_area, which controls the type returned by __getitem__, etc.
This lets us have a float32 response instead of float64 if desired.

Commit:
add .shape property for FullDataResponse analogous to Histogram

Commit:
Don't shuffle for compression when we are already bit-shuffling.
Empirically, we get slightly better compression and notably faster
decompression without the extra shuffle

Commit:
Rework .rsp to .h5 conversion so it does not keep two copies of the
entire response in memory in order to do F-to-C order transposition.
Instead, we do the transposition one chunk at a time, which roughly
halves memory usage for large responses.

Commit:
fix range too narrow error from line bg estimation with new response

Commit:
update generated test files from new responses

Commit:
avoid a copy when building spec/aeff responses

Commit:
Loading the full response to memory should return a DetectorResponse,
not a general Histogram.  Rename the relevant function from to_histogram()
to to_dr() and update its return type accordingly

Commit:
* Add accessor for raw counts dataset
* Add get_pixel() method to implement __getitem__,
  but with the possibility of specifying a weight that
  we can apply to eff_area rather than to the much larger
  counts matrix for greater efficiency.  Use this method
  in get_psr() and friends.

Commit:
- add tqdm progress bars to rsp-to-h5 conversion; may be disabled
  with new "quiet" option to RspConverter class
- response test cases now use get_pixel() rather than __getitem__
  for most tests

Commit:
Add "headers" property to FullDetectorResponse to access stored headers,
and use it in RspConverter when converting back to .rsp

jdbuhler committed:
Merge branch 'cositools:develop' into new_response_format

Commit:
update comment on compression with some new numbers

Commit:
small test tweaks to improve coverage of FullDetectorResponse.py

Commit:
* add ability to specify text buffer size for response converter,
  so we can exercise it with smaller sizes
* fix file exists check

jdbuhler committed:
Merge branch 'cositools:develop' into new_response_format

Commit:
Update newly added tutorial to use new-style response; a new
.h5 file was added to the develop directory on wasabi

Commit:
minor comment cleanup

israelmcmc commented:
Thanks again @jdbuhler!

I'll check this PR in detail in a couple of weeks. In the meanwhile, I'd like ask the following people to take a quick look at specific parts of the code: 
- @GallegoSav: can you please take a look at the RspConverter? I know some features have been dropped by Jeremy, I'm OK with that. 
- @hirokiyoneda: can you please check the code impacting the imaging deconvoltution module and the extended source setup in FullDetectorResponse?
- @eneights: can you please check the parts of the code that use the "Pol" axis? Also, can you double check that your ASAD results are still consistent. Jeremy changed that expected values for the unit test, which were using a very very coarse response, but I would expect that the results from a not-so-coarse response are still consistent.

None of these individual checks need to be in-dept, just a quick look for things the pop to your eyes would be helpful.

GallegoSav commented on cosipy/response/RspConverter.py:
```diff
@@ -0,0 +1,901 @@
+        'Dist': "Distance from first interaction"
+    }
+
+    # ordered subset of .rsp axes to keep for HDF5 response
```
is that important this order for hdf5 file ? Why not use the same as rsp ? It should not matter no ?

jdbuhler replied:
NuLambda being first is important pretty much everywhere, as rsp[i] is assumed to return a NuLambda slice.  In theory, we could change FullDetectorResponse.get_pixel() to accommodate arbitrary axis ordering, but for performance reasons, we might as well choose the order we are going to need later.  This is also the behavior of the previous conversion code.

jdbuhler replied:
I will add that the old code put PsiChi before Phi in the HDF5 axis order, but nothing else in cosipy assumes one or the other ordering of these two axes. Where a particular ordering *is* needed, Phi comes first, and the existing cosipy code does a reordering of each CDS in memory if needed to ensure this. So I figured the on-disk format might as well match what everyone actually uses.

jdbuhler replied:
One more note: I pushed a small change to verify that NuLambda is the first dimension of the response on disk when FullDetectorResponse opens the HDF5 file.  Removing this assumption would be a large undertaking, as many places in FullDetectorResponse assume that ._axes[1:] includes every dimension *except* NuLambda.  I don't believe there are any other such order dependencies in FullDetectorResponse besides this one.

GallegoSav commented on cosipy/response/RspConverter.py:
```diff
@@ -0,0 +1,901 @@
+        overwrite : bool
+           overwrite the target filename if it exists
+        compress: bool
+           true iff response HDF5 file should use internal
```
if

jdbuhler replied:
That's "iff" as in "if and only if".

GallegoSav commented on cosipy/response/RspConverter.py:
```diff
@@ -0,0 +1,901 @@
+
+            f.write(f"StartStream {counts.size}\n")
+
+            # counts are printed one one line in FORTRAN order
```
in one line

jdbuhler replied:
Will fix...

GallegoSav reviewed (commented):
Hi , I reviewed the code in the response folder. I just have minor comments about typo

Commit:
fix typo

Commit:
FullLDetectorResponse assumes that the on-disk response has NuLambda as
its first dimension.  Removing this assumption would be a large change with
performance implications for no clear benefit.  For now, verify the
assumption when we open the response file.

israelmcmc commented:
Thanks @GallegoSav!

eneights commented:
The code involving the polarization axis looks good to me!

For the fit performed in the ICRS frame, the ASAD method results with this new response format are consistent with the previous results. For the fit performed in the spacecraft frame, the fitted polarization fraction is slightly inconsistent but improved (80% +/- 6% vs 91% +/- 7%, where the injected value was 80%), with a slightly lower MDP than before.

jdbuhler commented:
Do we think the output change is because I replaced the test response?  I would not expect the code itself to cause a different result.

eneights commented:
By the test response, do you mean the coarse response used for the unit test?

My comment was about the ASAD method notebook, which uses the full resolution polarization response.

jdbuhler commented:
Ok. Not sure why that result should have changed... I will check on it.

jdbuhler commented:
I just tried the tutorial here, and I get identical results for the polarization spacecraft fit with the develop and new_response branches (and the corresponding response files, which I just re-downloaded from wasabi):

Best fit polarization fraction: 0.803 +/- 0.061
Minimum detectable polarization: 0.16
Best fit polarization angle (IAU convention): 87.419 +/- 2.058

jdbuhler commented:
Is there any chance this could be a dependency issue?  I am using

 - python 3.10.16
 - numpy 2.2.6
 - numba 0.61.2
 - scipy 1.13.0
 - astropy 6.1.7
 - healpy 1.18.1
 - h5py 1.18.1
 - hdf5plugin 5.1.0
 - histpy 2.0.3.dev0 (the develop branch)

jdbuhler commented:
I wonder if the saved output in the polarization tutorial's .ipynb file is out of date even for develop?  When I ran the develop branch's version of it, I did not get either the assorted graphs or the run of error messages, and even the ICRS result was not the same as what was stored.

Stored ICRS output:

Best fit polarization fraction: 0.953 +/- 0.073
Minimum detectable polarization: 0.164
Best fit polarization angle (IAU convention): 87.668 +/- 1.987

My ICRS output (same for develop and new_response branches):

Best fit polarization fraction: 0.936 +/- 0.072
Minimum detectable polarization: 0.161
Best fit polarization angle (IAU convention): 87.668 +/- 1.987

So perhaps the difference is due to other, prior changes to develop, and the tutorial output was just never updated?

eneights commented:
@jdbuhler yes, it's definitely possible that the notebook output was just not updated, and that's what I was comparing with (I didn't rerun with the old response version). If you're getting identical results, I think it's good to go from my side

israelmcmc commented:
I went through this in more detail and all looks good to me. I'll go ahead an merge it. Thanks @jdbuhler 

Some closing notes:
- After seeing the new `to_dr()` function and how it uses DectectorResponse, I think we can update the DectectorResponse docstring to specify that it also works with an optional NuLambda axis, not just specific pixel. But we can do that in a separate PR.
- @jdbuhler: I'm planning on closing #199, #110 and #106, and add a checkmark for the response module in the list in #226. If you have a chance, can you please take a look at those issues and confirm?
- Thanks @eneights and @GallegoSav for taking a look!
- @hirokiyoneda Feel free to still take a look, but I think we can address any comments on a follow up PR. After a more detailed look, I didn't see any major changes to the imaging module, and I know you are pretty busy with other stuff.

jdbuhler commented:
@israelmcmc I agree on your proposed actions for open issues.

israelmcmc commented:
Thanks @jdbuhler

hiyoneda commented:
@israelmcmc @jdbuhler Sorry for reviewing this PR so late. While it was closed, I checked that this change does not affect the image deconvolution modules. I ran through the notebooks for 511 keV image deconvolution for both galactic and scatt methods, and confirmed that the results are consistent with previous ones.

israelmcmc commented:
No worries, @hiyoneda! And thank you for confirming, that's good to hear.
